### DISCUSSION TITLE
How find tool algorithm works in OVE

### DISCUSSION THREAD

**Discussion by tnrich** [TeselaGen/openVectorEditor#853] [Q&A]
Copied from an email with @pmaen:

I am currently working on an open source bioinformatics tool and need to be able to search for sequences in circular plasmids. I have seen that your tool Open Vector Editor is capable of doing this, but I am not familiar with JavaScript. Could you perhaps explain to me what approach you used to implement this function?

Example of what I mean
<img width="384" alt="image" src="https://user-images.githubusercontent.com/2730609/182888102-e42cd3b1-da06-40c0-b4d3-5ccec919c9bb.png">

Basically, my problem is that when searching for a string in the plasmid, the string I am looking for could be, for example, the last 30 bases of the plasmid and the first 20 bases of the plasmid, which would be a 50 nt sequence on a circular plasmid, but I couldn't find a way to implement this in my tool.

Thank you in advance.

Kind regards
@pmaen

**Comment by tnrich** [comment]
Hi @pmaen, Here's where that code lives: 

<img width="1438" alt="image" src="https://user-images.githubusercontent.com/2730609/182888708-3d6c93aa-9c78-4ff4-a42f-605f873c4bb5.png">

The findSequenceMatches function: 
https://github.com/TeselaGen/ve-sequence-utils/blob/master/src/findSequenceMatches.js

Test code showing how the function is called:
https://github.com/TeselaGen/ve-sequence-utils/blob/master/src/findSequenceMatches.test.js#L110 

If you look in the code, you'll see the "trick" for finding matches that cross the "origin" of the sequence: 
<img width="457" alt="image" src="https://user-images.githubusercontent.com/2730609/182889246-78712516-1e1d-4a9b-bbf6-ecd3e15fb9e3.png">
Basically if the seq is circular, we duplicate the seq so we can find matches that span the origin. At the end of the function we then modulate any ranges that have ends that are "too long"  
<img width="534" alt="image" src="https://user-images.githubusercontent.com/2730609/182889597-8b764e47-21b0-4322-9bb2-ddb1ca12f40b.png">
as well as remove any duplicates caused by the doubling of the sequence:
<img width="253" alt="image" src="https://user-images.githubusercontent.com/2730609/182889713-fac404de-ee41-4b05-8558-fdcb944d2561.png">

Hope that all makes sense for you!
-Thomas

**Reply by pmaen** [reply]
Thank you very much for the thorough explanation. That helps a lot.

Kind regards
Paul
